### PR TITLE
polish_executor_and_add_ctx_cache

### PR DESCRIPTION
**Pull request by guru4elephant**
test=develop

### CODE REVIEW

**Comment by guru4elephant** [comment]
This PR mainly aims to speedup ExecutorPrepareContext creation. In particular, when a user set use_ctx_cache() in python API, ExecutorPrepareContext can be cached in Executor. This is useful for declarative programing that a graph is static and we do not need to modify the graph during training. 

I've tested the efficiency of this change on bert. The timeline before and after optimization are as follows. Note that this change can be generalized to other scenario such as resnet50, transformer and so on.

Before Optimization
image

After Optimization
image

In the future, ExecutorPrepareContext will be exposed to python API, and we can cache program and ExecutorPrepareContext together to make the cache strategy more clearly.

File: paddle/fluid/framework/executor.cc [hunk 1]
```diff
@@ -244,77 +252,21 @@ static bool has_fetch_operators(
...
-      op->SetOutput("Out", {var_name});
-      op->SetAttr("col", {static_cast<int>(i)});
-      op->CheckAttrs();
-
-      i++;
-    }
+void Executor::PrepareCtxCache(
+    const ProgramDesc& program, int block_id,
+    const std::vector<std::string>& skip_ref_cnt_vars, bool force_disable_gc) {
+  ctx_.reset(new ExecutorPrepareContext(program, block_id));
```
Comment (jacquesqiao) [inline paddle/fluid/framework/executor.cc]: add an mutex?

File: paddle/fluid/framework/executor.cc [hunk 2]
```diff
@@ -68,7 +63,11 @@ ExecutorPrepareContext::~ExecutorPrepareContext() {
   VLOG(5) << "destroy ExecutorPrepareContext";
 }
 
-Executor::Executor(const platform::Place& place) : place_(place) {}
+Executor::Executor(const platform::Place& place) : place_(place) {
+  cur_step_ = 0;
+  open_ctx_cache_ = false;
+  ctx_is_cached_ = false;
```
Comment (gavin1332) [inline paddle/fluid/framework/executor.cc]: could we do a nullptr checking of ctx_ to judge whether ctx_is_cached is false? And we should pay attention to caching multiple ctx

File: paddle/fluid/framework/executor.cc [hunk 3]
```diff
@@ -68,7 +63,11 @@ ExecutorPrepareContext::~ExecutorPrepareContext() {
   VLOG(5) << "destroy ExecutorPrepareContext";
 }
 
-Executor::Executor(const platform::Place& place) : place_(place) {}
+Executor::Executor(const platform::Place& place) : place_(place) {
+  cur_step_ = 0;
```
Comment (gavin1332) [inline paddle/fluid/framework/executor.cc]: what if i use this exe to run other programs, like startup program. Do we need a reseting method for context pamameters?

File: paddle/fluid/framework/executor.cc (hunk 1 above)
Comment (gavin1332) [reply paddle/fluid/framework/executor.cc]: why?

Comment (gavin1332) [reply paddle/fluid/framework/executor.cc]: > add an mutex?

why?

**Review by gavin1332 — DISMISSED** [review]
LGTM

**Review by jacquesqiao — APPROVED** [review]
LGTM!

**Comment by jacquesqiao** [comment]
Can you add a unit test to ensue that the cache works well.

**Pull request merged**